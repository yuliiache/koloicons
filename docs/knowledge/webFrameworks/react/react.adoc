== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







