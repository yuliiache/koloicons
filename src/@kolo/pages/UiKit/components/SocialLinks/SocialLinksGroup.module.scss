.socialLinksWrap {
  display: flex;
  width: fit-content;
}
