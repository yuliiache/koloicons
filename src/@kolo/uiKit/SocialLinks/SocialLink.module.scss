@import "src/@kolo/design";

.socialContainer {
  display: flex;
  gap: 0.5rem;
  justify-content: center;
  margin: 0;
  padding: 0;

  a {
    font-size: 0;

    &:focus {
      outline: none;
      box-shadow: 0 0 0 4px $IconFocus;
    }
  }

  svg {
    width: 1rem;
    height: 1rem;

    path,
    circle {
      fill: $SocialIcon;
      transition: fill 0.3s;
    }

    &:hover {
      path,
      circle {
        fill: $BrandColor;
      }
    }
  }
}
