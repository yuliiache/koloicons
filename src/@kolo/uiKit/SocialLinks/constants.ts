export enum SocialURL {
  INSTAGRAM = 'https://www.instagram.com/koloicons/',
  DRIBBLE = 'https://dribbble.com/Pictogramer',
}
