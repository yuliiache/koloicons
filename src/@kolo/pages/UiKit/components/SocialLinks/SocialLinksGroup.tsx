import SocialLinks from '@kolo/uiKit/SocialLinks/SocialLinks';

import {Section} from '../Section/Section';
import styles from './SocialLinksGroup.module.scss';

const SocialLinksGroup = () => {
  return (
    <Section title="SocialLinks">
      <div className={styles.socialLinksWrap}>
        <SocialLinks />
      </div>
    </Section>
  );
};

export default SocialLinksGroup;
