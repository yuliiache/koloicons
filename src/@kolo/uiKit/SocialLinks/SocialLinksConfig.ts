import React from 'react';

import Icon from '../Icons';
import {IconProps} from '../Icons/types/icon';
import {SocialURL} from './constants';

interface SocialLinkConfigItem {
  url: SocialURL;
  IconComponent: React.FC<IconProps>;
  id: string;
}

const SocialLinksConfig: SocialLinkConfigItem[] = [
  {
    url: SocialURL.INSTAGRAM,
    IconComponent: Icon.Instagram,
    id: 'social-links-instagram',
  },
  {
    url: SocialURL.DRIBBLE,
    IconComponent: Icon.Dribble,
    id: 'social-links-dribble',
  },
];

export default SocialLinksConfig;
