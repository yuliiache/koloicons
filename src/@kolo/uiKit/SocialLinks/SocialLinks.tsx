import {FC} from 'react';

import Link from '../Link/Link';
import styles from './SocialLink.module.scss';
import SocialLinksConfig from './SocialLinksConfig';

const SocialLinks: FC = () => {
  return (
    <div className={styles.socialContainer}>
      {SocialLinksConfig.map(({url, IconComponent, id}) => (
        <Link
          key={id}
          url={url}
          isInNewTab={true}
        >
          <IconComponent />
        </Link>
      ))}
    </div>
  );
};
export default SocialLinks;
