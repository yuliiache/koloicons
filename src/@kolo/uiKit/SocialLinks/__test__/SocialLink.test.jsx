import React from 'react';
import {cleanup, render, screen} from '@testing-library/react';

import '@testing-library/jest-dom';

import SocialLinks from '../SocialLinks';

describe('SocialLinks Component', () => {
  afterEach(() => {
    cleanup();
  });

  it('renders without crashing', () => {
    const {asFragment} = render(<SocialLinks />);
    expect(screen.getAllByRole('link')).toHaveLength(2);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders Instagram icon', () => {
    const {asFragment} = render(<SocialLinks />);
    expect(screen.getByTestId('icon-instagram')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders Dribble icon', () => {
    const {asFragment} = render(<SocialLinks />);
    expect(screen.getByTestId('icon-dribble')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('contains the correct Instagram link', () => {
    render(<SocialLinks />);
    const icon = screen.getByTestId('icon-instagram');
    const link = icon.closest('a');
    expect(link).toHaveAttribute('href', 'https://www.instagram.com/koloicons/');
  });

  it('contains the correct Dribble link', () => {
    render(<SocialLinks />);
    const icon = screen.getByTestId('icon-dribble');
    const link = icon.closest('a');
    expect(link).toHaveAttribute('href', 'https://dribbble.com/Pictogramer');
  });
});
