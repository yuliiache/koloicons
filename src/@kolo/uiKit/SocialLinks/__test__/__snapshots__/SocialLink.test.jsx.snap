// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialLinks Component renders Dribble icon 1`] = `
<DocumentFragment>
  <div
    class="socialContainer"
  >
    <a
      class="link link--default"
      data-testid="externalLinkComponent"
      href="https://www.instagram.com/koloicons/"
      rel="noopener noreferrer nofollow"
      target="_blank"
    >
      <svg
        data-testid="icon-instagram"
        enable-background="new 0 0 60 60"
        height="20"
        viewBox="0 0 60 60"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="30"
          cy="30"
          r="10"
        />
        <path
          d="m49 6h-38c-2.761 0-5 2.239-5 5v38c0 2.761 2.239 5 5 5h38c2.761 0 5-2.239 5-5v-38c0-2.761-2.239-5-5-5zm-19 39c-8.284 0-15-6.716-15-15s6.716-15 15-15 15 6.716 15 15-6.716 15-15 15zm16-28c-1.657 0-3-1.343-3-3s1.343-3 3-3 3 1.343 3 3-1.343 3-3 3z"
        />
      </svg>
    </a>
    <a
      class="link link--default"
      data-testid="externalLinkComponent"
      href="https://dribbble.com/Pictogramer"
      rel="noopener noreferrer nofollow"
      target="_blank"
    >
      <svg
        data-testid="icon-dribble"
        enable-background="new 0 0 60 60"
        height="20"
        viewBox="0 0 60 60"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m34.706 24.503c.814 1.79 1.594 3.595 2.341 5.413 2.906-.599 5.902-.916 8.953-.916 3.059 0 6.063.316 8.998.919-.019-5.69-1.94-10.93-5.158-15.123-4.901 4.275-9.839 7.454-15.134 9.707z"
        />
        <path
          d="m33.131 30.911c-.695-1.675-1.418-3.338-2.172-4.987-6.39 2.137-13.37 3.076-21.459 3.076-1.494 0-2.985-.051-4.471-.149-.017.381-.029.764-.029 1.149 0 6.155 2.228 11.786 5.916 16.142 5.636-7.202 13.401-12.557 22.215-15.231z"
        />
        <path
          d="m46 33c-2.535 0-5.025.245-7.452.701 2.125 5.569 3.951 11.251 5.438 17.023 5.615-3.796 9.601-9.815 10.704-16.792-2.83-.618-5.732-.932-8.69-.932z"
        />
        <path
          d="m9.5 25c7.403 0 13.823-.843 19.712-2.765-2.584-5.285-5.465-10.426-8.635-15.392-7.603 3.097-13.339 9.824-15.049 18.033 1.32.082 2.645.124 3.972.124z"
        />
        <path
          d="m32.974 20.825c4.951-2.093 9.574-5.044 14.165-9.016-4.476-4.219-10.503-6.809-17.139-6.809-1.876 0-3.701.213-5.459.605 3.092 4.917 5.898 10.004 8.433 15.22z"
        />
        <path
          d="m34.626 34.638c-8.324 2.445-15.643 7.513-20.866 14.361 4.369 3.738 10.039 6.001 16.24 6.001 3.698 0 7.205-.809 10.363-2.25-1.539-6.148-3.456-12.198-5.737-18.112z"
        />
      </svg>
    </a>
  </div>
</DocumentFragment>
`;

exports[`SocialLinks Component renders Instagram icon 1`] = `
<DocumentFragment>
  <div
    class="socialContainer"
  >
    <a
      class="link link--default"
      data-testid="externalLinkComponent"
      href="https://www.instagram.com/koloicons/"
      rel="noopener noreferrer nofollow"
      target="_blank"
    >
      <svg
        data-testid="icon-instagram"
        enable-background="new 0 0 60 60"
        height="20"
        viewBox="0 0 60 60"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="30"
          cy="30"
          r="10"
        />
        <path
          d="m49 6h-38c-2.761 0-5 2.239-5 5v38c0 2.761 2.239 5 5 5h38c2.761 0 5-2.239 5-5v-38c0-2.761-2.239-5-5-5zm-19 39c-8.284 0-15-6.716-15-15s6.716-15 15-15 15 6.716 15 15-6.716 15-15 15zm16-28c-1.657 0-3-1.343-3-3s1.343-3 3-3 3 1.343 3 3-1.343 3-3 3z"
        />
      </svg>
    </a>
    <a
      class="link link--default"
      data-testid="externalLinkComponent"
      href="https://dribbble.com/Pictogramer"
      rel="noopener noreferrer nofollow"
      target="_blank"
    >
      <svg
        data-testid="icon-dribble"
        enable-background="new 0 0 60 60"
        height="20"
        viewBox="0 0 60 60"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m34.706 24.503c.814 1.79 1.594 3.595 2.341 5.413 2.906-.599 5.902-.916 8.953-.916 3.059 0 6.063.316 8.998.919-.019-5.69-1.94-10.93-5.158-15.123-4.901 4.275-9.839 7.454-15.134 9.707z"
        />
        <path
          d="m33.131 30.911c-.695-1.675-1.418-3.338-2.172-4.987-6.39 2.137-13.37 3.076-21.459 3.076-1.494 0-2.985-.051-4.471-.149-.017.381-.029.764-.029 1.149 0 6.155 2.228 11.786 5.916 16.142 5.636-7.202 13.401-12.557 22.215-15.231z"
        />
        <path
          d="m46 33c-2.535 0-5.025.245-7.452.701 2.125 5.569 3.951 11.251 5.438 17.023 5.615-3.796 9.601-9.815 10.704-16.792-2.83-.618-5.732-.932-8.69-.932z"
        />
        <path
          d="m9.5 25c7.403 0 13.823-.843 19.712-2.765-2.584-5.285-5.465-10.426-8.635-15.392-7.603 3.097-13.339 9.824-15.049 18.033 1.32.082 2.645.124 3.972.124z"
        />
        <path
          d="m32.974 20.825c4.951-2.093 9.574-5.044 14.165-9.016-4.476-4.219-10.503-6.809-17.139-6.809-1.876 0-3.701.213-5.459.605 3.092 4.917 5.898 10.004 8.433 15.22z"
        />
        <path
          d="m34.626 34.638c-8.324 2.445-15.643 7.513-20.866 14.361 4.369 3.738 10.039 6.001 16.24 6.001 3.698 0 7.205-.809 10.363-2.25-1.539-6.148-3.456-12.198-5.737-18.112z"
        />
      </svg>
    </a>
  </div>
</DocumentFragment>
`;

exports[`SocialLinks Component renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="socialContainer"
  >
    <a
      class="link link--default"
      data-testid="externalLinkComponent"
      href="https://www.instagram.com/koloicons/"
      rel="noopener noreferrer nofollow"
      target="_blank"
    >
      <svg
        data-testid="icon-instagram"
        enable-background="new 0 0 60 60"
        height="20"
        viewBox="0 0 60 60"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="30"
          cy="30"
          r="10"
        />
        <path
          d="m49 6h-38c-2.761 0-5 2.239-5 5v38c0 2.761 2.239 5 5 5h38c2.761 0 5-2.239 5-5v-38c0-2.761-2.239-5-5-5zm-19 39c-8.284 0-15-6.716-15-15s6.716-15 15-15 15 6.716 15 15-6.716 15-15 15zm16-28c-1.657 0-3-1.343-3-3s1.343-3 3-3 3 1.343 3 3-1.343 3-3 3z"
        />
      </svg>
    </a>
    <a
      class="link link--default"
      data-testid="externalLinkComponent"
      href="https://dribbble.com/Pictogramer"
      rel="noopener noreferrer nofollow"
      target="_blank"
    >
      <svg
        data-testid="icon-dribble"
        enable-background="new 0 0 60 60"
        height="20"
        viewBox="0 0 60 60"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m34.706 24.503c.814 1.79 1.594 3.595 2.341 5.413 2.906-.599 5.902-.916 8.953-.916 3.059 0 6.063.316 8.998.919-.019-5.69-1.94-10.93-5.158-15.123-4.901 4.275-9.839 7.454-15.134 9.707z"
        />
        <path
          d="m33.131 30.911c-.695-1.675-1.418-3.338-2.172-4.987-6.39 2.137-13.37 3.076-21.459 3.076-1.494 0-2.985-.051-4.471-.149-.017.381-.029.764-.029 1.149 0 6.155 2.228 11.786 5.916 16.142 5.636-7.202 13.401-12.557 22.215-15.231z"
        />
        <path
          d="m46 33c-2.535 0-5.025.245-7.452.701 2.125 5.569 3.951 11.251 5.438 17.023 5.615-3.796 9.601-9.815 10.704-16.792-2.83-.618-5.732-.932-8.69-.932z"
        />
        <path
          d="m9.5 25c7.403 0 13.823-.843 19.712-2.765-2.584-5.285-5.465-10.426-8.635-15.392-7.603 3.097-13.339 9.824-15.049 18.033 1.32.082 2.645.124 3.972.124z"
        />
        <path
          d="m32.974 20.825c4.951-2.093 9.574-5.044 14.165-9.016-4.476-4.219-10.503-6.809-17.139-6.809-1.876 0-3.701.213-5.459.605 3.092 4.917 5.898 10.004 8.433 15.22z"
        />
        <path
          d="m34.626 34.638c-8.324 2.445-15.643 7.513-20.866 14.361 4.369 3.738 10.039 6.001 16.24 6.001 3.698 0 7.205-.809 10.363-2.25-1.539-6.148-3.456-12.198-5.737-18.112z"
        />
      </svg>
    </a>
  </div>
</DocumentFragment>
`;
